gh-264 Base changes to support theming of application

pjmonks opened the pull request:
- Rename existing Angular Material theme from "default-theme" to "active-theme" - files, classes and references
- Create ThemeService to manage loading and applying themes to the DOM. Loading is hardcoded for now using existing colour scheme
- Generate Material palette colours for application using only 3 base colours
- Fix typography settings for theme and make customisable
- Update SCSS stylesheet to use latest Angular Material mixins
- Add back the missing typography settings to the theme, default to the existing values
- Rename existing Sass colour variables
- Add colour references to components that are not Angular Material
- Fix _resets.scss because Sass and CSS3 variables were incompatible. Darken colour value using HSL instead
- Change colour of data-class-row component to fit with the default colour scheme better
- Add request status colours to theme
- Add request counter theme colour
- Tests to ensure that CSS variables are set correctly
- Create documentation page for theming

**Note:** requires a new dependency on the `tinycolor2` library, please run `npm install` before testing.

Resolves #264 

# Screenshots

Default, existing theme:

![image](https://user-images.githubusercontent.com/3219480/220108981-021dd7cb-acf1-4579-979d-cd0963e677de.png)

Testing modifying the primary colour by temporarily altering the `defaultTheme.material.colors.primary` field:

![image](https://user-images.githubusercontent.com/3219480/220110192-c1383f39-c013-40c3-9113-1f75da2daa7d.png)

jamesrwelch commented:
Looking good 👍

jamesrwelch commented:
(That wasn't an official review)

NigelPalmer commented:
@pjmonks , I can't get this to build. I keep getting this error:

Error: src/app/shared/theme.service.ts:21:23 - error TS2307: Cannot find module 'tinycolor2' or its corresponding type declarations.

21 import tinycolor from 'tinycolor2';

pjmonks commented:
> @pjmonks , I can't get this to build. I keep getting this error:
> 
> Error: src/app/shared/theme.service.ts:21:23 - error TS2307: Cannot find module 'tinycolor2' or its corresponding type declarations.
> 
> 21 import tinycolor from 'tinycolor2';

I've updated the PR description to state you need to run `npm install` first.

NigelPalmer reviewed (changes requested):
Overall this is looking good and it allows for colours to be easily changed. But, there are a few problems.

### Default Theme Issues
The default colours that are now being used are not the same as they were before. There are a number of discrepancies between the old theming and the new theming. I only checked the Green, but I expect the gold and red schemes to also have discrepancies. The left column are the original colours, and the right column are the new calculated colours.
![Pallette Differences](https://user-images.githubusercontent.com/110250803/220157616-609a3a8b-cb5b-4cd1-ba99-f64079b1c2cd.JPG)

The colour of the class header has changed from grey to gold:

Original:
![classColourDev](https://user-images.githubusercontent.com/110250803/220157895-dc2f1a42-5837-4ae4-98fd-0a2e2db02265.JPG)

New:
![classColourThemed](https://user-images.githubusercontent.com/110250803/220157948-00959b38-a428-4988-bb23-eb0012181dad.JPG)

The progress bar has changed from Green to Grey.

Original:
![NoElementsDev](https://user-images.githubusercontent.com/110250803/220158167-86a664ed-e8b5-4f8c-9141-ffcabc7f381c.JPG)

New:
![NoElementsThemed](https://user-images.githubusercontent.com/110250803/220158231-392f18d5-aea8-4d5f-ba38-f4c57e165e77.JPG)

The unsent badge now has black text. (although this is probably the correct contrast colour)
![unsentColorChanged](https://user-images.githubusercontent.com/110250803/220159420-6127b3d7-c669-4c0d-a448-c97383e35724.JPG)


### Custom Theme Issues
The $color-white is used in a lot of css files still. I think I overlooked this when specifying the task. The knock on effect is that text is not being contrasted properly if a light custom colour is chosen. To improve the theming we either have to choose the colour automatically or allow the user to set it.
![whiteText](https://user-images.githubusercontent.com/110250803/220159289-08c13ed9-6972-47bd-83a4-8e221faec106.JPG)

pjmonks commented on src/app/shared/header/header.component.scss:
```diff
@@ -23,8 +23,8 @@ $link-color: #ffffff;
-$header-background: $color-mauro-dark-green;
-$request-count-color: $color-yellow;
+$header-background: $color-primary;
+$request-count-color: $color-requestCount; //$color-yellow;
```
Removed

pjmonks commented on src/styles/base/_colours.scss:
```diff
@@ -27,23 +27,53 @@ $color-gray-90: #e6e6e6;
 $color-gray-85: #d7d7d7;
 $color-gray-75: #c4c4c4;
 $color-gray-38: #606060;
```
Removed the gray values which are not used anymore

pjmonks commented:
Updated with changes requested from feedback:

- Code cleanup
- Unsent and submitted badge colours have been adjusted
- Fixed Material colour palette generation, was using wrong algorithm before. This also fixes the progress bar colours
- Updated other components to have correct colour contrast of text when a theme colour is used as a background.
- Added new theme colour "classRow" to put back the grey background for `data-class-row` component

Note that to support wider theming there may be additional minor tweaks to the UI.

NigelPalmer reviewed (approved):
This all looks good now. All my complaints have been dealt with and it is easy to change the colours of the website. A good job, well done! :thumbsup:
